### PR TITLE
[firebase_auth] Adds  a Stream of FirebaseUser that emits on each user update (#717).

### PR DESCRIPTION
**Pull request by feinstein**
## Description

I added a second `Stream` to the `FirebaseAuth` class that emits a new `FirebaseUser` at everytime there's a new sign-in or sign-out (same as `onAuthStateChanged`), but also emits if there's a call to `currentUser()` and the `FirebaseUser` returned is different than the last one.

This is expected to be used as:

```dart
await _user.reload(); // needs to call reload to get new data on the server.
_user = await _auth.currentUser(); // triggers the loading of the new data and updates onUserUpdated.
```

## Related Issues

#717 

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] If the pull request affects only one plugin, the PR title starts with the name of the plugin in brackets (e.g. [cloud_firestore])
- [ ] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [ ] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/FirebaseExtended/flutterfire/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

**Comment by feinstein** [comment]
I tried to make unit tests, but I couldn't find a way to make relevant tests, as the `FirebaseAuth` class is heavily mocked and `FirebaseUser` only has private constructors. I will need help from someone more experienced than me in Dart to produce any relevant tests on this feature.

**Comment by feinstein** [comment]
@collinjackson could you take a look at this?

File: packages/firebase_auth/firebase_auth/lib/src/firebase_auth.dart [hunk 1]
```diff
@@ -18,13 +18,36 @@ class FirebaseAuth {
...
+  /// Receive [FirebaseUser] each time [onAuthStateChanged] emits an event, or
+  /// when a call to [currentUser] returns a different [FirebaseUser].
+  ///
+  /// Expected to be used like:
+  ///
+  /// ```dart
+  /// ...
+  /// // relaods data from the server, like isEmailVerified
+  /// user.reload();
+  /// // relaods teh actual user with the new data and triggers the stream
```
Comment (collinjackson) [inline packages/firebase_auth/firebase_auth/lib/src/firebase_auth.dart]: typo here

File: packages/firebase_auth/firebase_auth/lib/src/firebase_auth.dart [hunk 2]
```diff
@@ -18,13 +18,36 @@ class FirebaseAuth {
...
   }
 
+  /// Receive [FirebaseUser] each time [onAuthStateChanged] emits an event, or
+  /// when a call to [currentUser] returns a different [FirebaseUser].
+  ///
+  /// Expected to be used like:
+  ///
+  /// ```dart
+  /// ...
+  /// // relaods data from the server, like isEmailVerified
```
Comment (collinjackson) [inline packages/firebase_auth/firebase_auth/lib/src/firebase_auth.dart]: typo here

**Comment by feinstein** [comment]
I understand that strategy, but at the same time Flutter is highly reactive, whereas Android and iOS are not, so IMO the API should match the native ones, but extensions that would make it work better on Flutter are desirable, as long as they don't interfere with the original native API.

**Comment by feinstein** [comment]
Since this won't be merged, I made a package that implement some of this functionality and more: [firebase_user_stream](https://pub.dev/packages/firebase_user_stream)

**Pull request closed without merging**